### ISSUE TITLE
With/Without examples

### ISSUE THREAD

**Issue by Extrawurst** [Ralith/hecs#33]
as a followup to https://github.com/Ralith/hecs/pull/32 I think I am not the only one who would benefit from having more complete examples of how to combine `With` and `Without` with actual component data in `query` now that the `with()` and `without()` chaining helpers are gone.

**Comment by Ralith** [comment]
I believe the examples on `With`/`Without` themselves satisfy this request, but that's definitely not optimal for discoverability. Does #34 address your concern?

I'm open to restoring the combinator methods if they seem substantially more ergonomic. They definitely have an advantage in that they don't require anything to be imported.

**Comment by Extrawurst** [comment]
Hi @Ralith first of let me say I appreciate you taking the time for me here, I love the approach you take hecs in terms of simplicity compared to other rust ECS's.

Regarding the issue at hand, I wrote a little comparison of the `With`/`Without` approach vs. `with()`/`without()`:
```rust
use hecs::{With, Without, World};

struct Tag1;
struct Tag2;

fn main() {
    let mut world = World::new();

    let a = world.spawn((123, Tag1 {}, true));
    let b = world.spawn((42, Tag2 {}, true));
    let c = world.spawn((1, true));

    for (_, number) in &mut world.query::<With<bool, Without<Tag1, Without<Tag2, &mut i32>>>>() {
        *number += 1;
    }
    assert_eq!(*world.get::<i32>(a).unwrap(), 123);
    assert_eq!(*world.get::<i32>(b).unwrap(), 42);
    assert_eq!(*world.get::<i32>(c).unwrap(), 2);

    for (_, number) in &mut world
        .query::<&mut i32>()
        .with::<bool>()
        .without::<Tag1>()
        .without::<Tag2>()
    {
        *number += 1;
    }
    assert_eq!(*world.get::<i32>(a).unwrap(), 123);
    assert_eq!(*world.get::<i32>(b).unwrap(), 42);
    assert_eq!(*world.get::<i32>(c).unwrap(), 3);
}
```

Note that this is a already more simplified version of more complex queries in my code base.
1) the chaining approach is easier to read in my opinion
2) you are right, the lack of having to import `use` the types is a plus aswell
3) perhaps my most important factor: It is much more obvious what my query *actually* cares about (the i32) and separates that more clearly from the with/without stuff that is wrapped up around it in the other case.
4) these more complex but also more real life examples were not obvious to me when I read the `With`/`Without` docs. So no #34 would not cut it in my opinion.

*TL;DR;*
After these considerations I am clearly in favour of keeping the chaining combinators and actually even considering making them the prime approach to this instead of the types. Maybe even make the Types private and undoc'd

**Comment by Ralith** [comment]
I'm glad you're finding hecs useful, and thanks for your analysis! I agree that this makes a compelling case for preserving the combinators.
